### ISSUE TITLE
Implicit inline classes with virtual functions (breaks libraries depending on inet when using opp 5.0)

### ISSUE THREAD

**Issue by tillsteinbach** [inet-framework/inet#191]
Hi all,
there are classes in inet that are implicitly inline (no compilation unit). With the new gcc 5 in OMNeT++ 5.0 this seems to breaks libraries depending on inet:

`
clang++.exe: warning: argument unused during compilation: '-shared-libgcc'
.tmplib4236(SRPRelay.o):(.rdata[__ZTIN4inet14Ieee8021dRelayE]+0x18): undefined reference to `typeinfo for inet::ILifecycle'
.tmplib4236(BGEtherLLC.o):(.rdata[__ZTIN4inet8EtherLLCE]+0x18): undefined reference to `typeinfo for inet::ILifecycle'
.tmplib4236(Shapers.o):(.rdata[__ZTIN4inet7MACBaseE]+0x18): undefined reference to `typeinfo for inet::ILifecycle'
.tmplib4236(BaseShaper.o):(.rdata+0x808): undefined reference to `typeinfo for inet::IPassiveQueue'
.tmplib4236(IPv4oREBase.o):(.rdata[__ZTIN4inet4IPv4E]+0x18): undefined reference to `typeinfo for inet::INetfilter'
.tmplib4236(IPv4oREBase.o):(.rdata[__ZTIN4inet4IPv4E]+0x20): undefined reference to `typeinfo for inet::ILifecycle'
.tmplib4236(IPv4oREBase.o):(.rdata[__ZTIN4inet4IPv4E]+0x28): undefined reference to `typeinfo for inet::INetworkProtocol'
.tmplib4236(TrafficPattern.o): In function `ZN9CoRE4INET14TrafficPattern7matchesEPKN7omnetpp7cPacketE':
C:\Users\muste\Documents\git\CoRE\CoRE4INET\src/core4inet/networklayer/inet/base/TrafficPattern.cc:67: undefined reference to `_imp___ZTIN4inet9UDPPacketE'
.tmplib4236(TrafficPattern.o):(.rdata[__ZTIN4inet12IPv4DatagramE]+0x18): undefined reference to `typeinfo for inet::INetworkDatagram'
.tmplib4236(TrafficPattern.o):(.rdata[__ZTIN4inet3tcp10TCPSegmentE]+0x18): undefined reference to `typeinfo for inet::ITransportPacket'
.tmplib4236(SRProtocol.o):(.rdata[__ZTIN4inet11Ieee802CtrlE]+0x18): undefined reference to `typeinfo for inet::IMACProtocolControlInfo'
clang++.exe: error: linker command failed with exit code 1 (use -v to see invocation)
`
I can fix that by just moving one of the methods to a compilation unit.

In general there are lots of classes where we have methods that are declared virtual but are implemented in the header (and are thus implicitly inline) I'm not a c++ expert, but that looks like trouble! Where does the vtable go for those methods? 
Another question to talk about is: Does it make sense to declare getters and setters virtual!?

I'm afraid, it looks like we have some cleanup coming up, but before we start this should be discussed, right?

**Comment by tillsteinbach** [comment]
Tested this with gcc 5.3 on debian and it seems to be only a windows problem. I guess it is again related to the dll-export keyword. My library tries to import from the shared lib but the code is not there as there is no compilation unit. Same problem as the template classes

**Comment by tillsteinbach** [comment]
It looks like the INET_API macro is again the problem. I think the problem is:

- the class is completely inline thus there is no virtual table being exported (usually the table would be in the compilation unit of the first function that is defined outline.
- Now I import INET and my compiler finds the class with dllimport and thus does not export the symbol itself because the dllimport suggests that it is available somewhere (usually it would export the symbol and let the linker then delete duplicates)

This is my guess. If I import the files after removing the INET_API define it works. But that is not a good way to go.
So I would really vote that INET should not contain any class that does not have an outline defined virtual method (at least a trivial virtual destructor can be defined in the compilation unit). So how to find the classes where this is wrong: clang provides a warning -Wweak-vtables

With clang being the default compiler on windows soon, we should work that out fast!

Best regards
Till

**Comment by rhornig** [comment]
I've turned on the -Wweak-vtables but it give warnings also for all the interfaces we use (i.e. where all methods are pure virtual). Having an outline definition for those would pretty much defy the purpose

**Comment by tillsteinbach** [comment]
Well, I could live with those having no outline definition, but shouldn't those have at least a default virtual destructor that can be outline defined?

**Comment by rhornig** [comment]
Oh my... After hunting this for the better part of the day... It has totally nothing to do with the above. I should have figured out earlier, but I've just realized: The whole INET_API macro stuff is JUST for the MS Visual C++ compiler (i.e. where you explicitly have mark with export/import all functions/class wwhat you want to use). Other compilers on Windows (GCC/CLANG) can run in auto export/import mode where all the public symbols are exported automatically. So, on these compilers the INET_API macro should be expanded to empty and they will do their stuff just fine...

Now unfortunately the main makefile in INET contains a line

`MAKEMAKE_OPTIONS := -f --deep -o INET -O out -pINET --no-deep-includes -I.`

The -pINET indicates that the makefile should pass the INET_EXPORT macro to the compiler, actually triggering the export macro on all compilers. Now it was not causing any issues on non-windows platforms as it expands to OPP_EXPORT which expands to empty on all but on Windows platform. It seems also that MinGW simply ignores `__declspec(dllexport)`, so that also works in Windows. 

Clang on the other hand tries to emulate the behavior of the Visual C++ compiler, and actually understands the `__declspec(dllexport)` marker, but it still cannot properly import the stuff...

Long story short, the `-pINET` option should be removed from the Makefile (and from the IDE project file) and the whole makefile must be regenerated (i.e. make cleanall; make makefiles)

After this the auto import/export will function correctly on clang too.

Let me know if that fixes the issue.

**Comment by tillsteinbach** [comment]
Oh, and I thought it was intended that __declspec(dllexport) is set to generate somehow compatible dlls. That will prevent a lot of problems in the future. dllexport was causing a lot of problems recently.

I just checked with our windows jenkins slave and it seems to work when removing -pINET.

My question is: wouldn't it be better to leave the -pINET and change the macro in platdefs.h to check for (_WIN32 and not __clang__) or best to change the _WIN32 check to a check for _MSC_VER?

**Comment by rhornig** [comment]
Yep. I strict sense, checking for _MSC_VER would be more appropriate, however that code is inside OMNeT++ and that ship has sailed even for the 5.0 version, not to mention for all released 4.x versions :) Moreover who knows, clang is indeed aiming to be binary compatible with Visual C on long term, especially that Microsoft is shipping the clang/c2 toolchain, so probably on long term, we will indeed see a clang version which properly supports export/import and generates an ABI compatible version for VC.

On the other hand, removing the -pINET should not hurt anyone, because that generates a GNU makefile which is usable only under macOS, Linux and MinGW. That makefile cannot work with Visual C anyway so it's not a problem that we omit the INET_EXPORT macro...

**Comment by tillsteinbach** [comment]
OK! Then lets go that way, I'll remove all -DINET_IMPORT from our projects.

**Comment by rhornig** [comment]
Ok, I've removed the -pINET from the GNU makefile

**Comment by tillsteinbach** [comment]
Will there be a patch release for INET soon? Otherwise I will have to prepare a small patch with this changes for my users.

**Comment by rhornig** [comment]
We plan to release a INET 3.3 version with that change next week.

**Comment by tillsteinbach** [comment]
What about the 3.3 release? An intermediate 3.2.5 would be enough for me :)

**Comment by rhornig** [comment]
We are currently merging all changes. So probably it does not make sense to release a 3.2.5 version. We still have to merge in two branches before relase.

**Comment by rhornig** [comment]
@tillsteinbach I'm right now working towards INET 3.3 release. If you can would you be able to check the mater branch whether that will be ok for your projects? INET 3.3 will be hopefully the last version that still runs with OMNeT++ 4.6.

**Comment by tillsteinbach** [comment]
Sure, I try to check that

**Comment by tillsteinbach** [comment]
@rhornig Did you by mistake remove "--meta:export-library" in .oppbuildspec? This should be definitely set, otherwise my dependent libraries won't link against INET!

**Comment by rhornig** [comment]
@tillsteinbach You are right. I will correct that. Thanks.

**Comment by rhornig** [comment]
@tillsteinbach Corrected the issue in master

**Comment by tillsteinbach** [comment]
Thanks! Sorry for finding that late, this didn't pop up in our jenkins tests as we cannot use the IDE to build there.

**Comment by tillsteinbach** [comment]
Ok, we checked on Windows, Debian, and OSX and it looks good to go now

Best regards
till

**Comment by AlharthiAmeera** [comment]
hi , I have same problem in my project . Please tell me how can i sove it >> My OS windows7 omnet5.2.1 inet inet-3.6.4-394571f

**Comment by rhornig** [comment]
This patch should workaround this issue: https://github.com/inet-framework/inet/commit/6937f43e5f94216d022fdf7556cbd4c29e256ffd
